src/app/loadout/ModPicker.tsx

```diff
@@ -334,35 +344,60 @@ function isModSelectable(
   if (selected.includes(mod) && unstackableModHashes.includes(mod.hash)) {
     return false;
   }
 
+  const { plugCategoryHash } = mod.plug;
+
+  const isSlotSpecificCategory = slotSpecificPlugCategoryHashes.includes(plugCategoryHash);
+
   // If there's an already selected mod that excludes this mod, we can't select this one
   const exclusionGroup = getModExclusionGroup(mod);
   if (exclusionGroup && selected.some((mod) => getModExclusionGroup(mod) === exclusionGroup)) {
     return false;
   }
 
+  const maximumPotentiallyAvailableArmorSlotEnergyCapacities =
+    getMaximumPotentiallyAvailableArmorSlotEnergyCapacities(selected);
+
   if (isSlotSpecificCategory) {
-    // General and slot-specific mods just match to the same category hash
-    const associatedLockedMods = selected.filter(
-      (mod) => mod.plug.plugCategoryHash === plugCategoryHash,
-    );
-    // Slot-specific mods (e.g. chest mods) can slot 3 per piece, so make sure the sum of energy doesn't
-    // exceed the maximum and that energy all aligns. This doesn't check other mods that could be on the
-    // item because we haven't assigned those to specific pieces.
-    const lockedModCost = isSlotSpecificCategory
-      ? _.sumBy(associatedLockedMods, (mod) => mod.plug.energyCost?.energyCost || 0)
-      : 0;
-    const modCost = energyCost?.energyCost || 0;
-
-    return lockedModCost + modCost <= MAX_ARMOR_ENERGY_CAPACITY;
+    const bucketHash = getBucketHashFromPlugCategoryHash(plugCategoryHash);
+    if (bucketHash === undefined) {
+      // This should never happen, but if it does, we don't want to crash the app
+      errorLog('loadout', 'bucketHash is undefined', mod);
+      return false;
+    }
+
+    const modCost = getModCost(mod);
+    if (modCost > maximumPotentiallyAvailableArmorSlotEnergyCapacities[bucketHash]) {
+      return false;
+    }
   }
 
+  const {
+    modMap: { generalMods, activityMods },
+  } = categorizeArmorMods(selected);
+  // General mods (Stat boosting mods) and Activity mods (Raid, Nightmare) are armor slot agnostic
+  // We need to check both of these groups separately
+  // Note that Artifice mods are also slot agnostic, but they have no cost so they don't need to be checked
+  for (const armorSlotAgnosticModGroup of [generalMods, activityMods]) {
+    // Can the highest cost mod fit in the highest remaining armor energy slot?
+    // Can the second highest cost mod fit in the second highest remaining armor energy slot?
+    // etc...
+    const potentialSelectedModCosts = padCostArray(
+      [mod, ...armorSlotAgnosticModGroup].map((mod) => getModCost(mod)),
+    ).sort((a, b) => b - a);
+    const availableEnergy = Object.values(
+      maximumPotentiallyAvailableArmorSlotEnergyCapacities,
+    ).sort((a, b) => b - a);
+    const hasValidModPlacement = potentialSelectedModCosts.every(
+      (modCost, index) => modCost <= availableEnergy[index],
+    );
+    if (!hasValidModPlacement) {
+      return false;
+    }
+  }
```

Comment:
This logic seems dubious/not correctly commented to me - `maximumPotentiallyAvailableArmorSlotEnergyCapacities` already includes the costs from `generalMods`/`activityMods`, so I don't understand why we're checking generalMods/activityMods against the optimistic remeaining energy *again*.

Also `armorSlotAgnosticModGroup` may have 5 elements so that `[mod, ...armorSlotAgnosticModGroup]` has 6 elements (but gets clamped down to 5 by `padCostArray` again), which makes it hard to verify that this is working as intended.

Reply:
Oh wow. Yeah this logic is totally wrong on re-reading it. It should be much simpler.

Reply:
Ok I have re-written this logic.